fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,383 @@
+/*
+ * FDBDirectory.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.directory;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.LockFactory;
+import org.apache.lucene.store.NoLockFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Directory implementation backed by FDB which attempts to
+ * model a file system on top of FoundationDB.
+ *
+ * @see <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/store/Directory.html">Directory</a>
+ */
+@API(API.Status.EXPERIMENTAL)
```

Comment:
I think it would be good to have a brief coverage of the transactional semantics of `FDBDirectory` and the related objects. Specifically, it would be good to have a sense of when `FDBDirectory` objects should be created, and how they should handle their transactions.

From reading the code, it seems that it implicitly is assuming that an `FDBDirectory` is scoped to a single fdb `Transaction`, which implies (to me, anyway) that this object won't be useful or correct once that transaction has been terminated. 

If that is the case, then it would be good to note that as it opens up possible programming errors from badly behaved callers. It also opens up some questions here about the overall usefulness of having `fileReferenceCache` and `blockCache` be transaction-scoped caches; questions like "are we using the correct amount of memory for those caches? Do they provide significant benefit when scoped down that narrowly?" and so on. 

If that's not the case (e.g. if it's assumed that a `FDBDirectory` is a long-lived object), then I'd be worried about the semantics of holding a single mutable transaction field here, and whether the code as written is correct when using a `FDBDirectory` object across multiple transactions.

Reply:
You are correct that the directory is smoked once a transaction is committed or rolled back.  The caching problem is interesting.  Those caches were added do the chattiness of the Lucene protocol.  FDBDirectory by definition cannot be held for more than 5 seconds currently.